File: C++/Letter_Combinations_of_a_Phone_Number.cpp
//LeetCode Problem link - https://leetcode.com/problems/letter-combinations-of-a-phone-number/

class Solution {
public:
    vector<char> getCharsForDigit(int x)
    {
        vector<char> dummy;
        if(x==2)
        {
            vector<char> c;
            c.push_back('a');
            c.push_back('b');
            c.push_back('c');
            return c;
        }
        else if(x==3)
        {
            vector<char> c;
            c.push_back('d');
            c.push_back('e');
            c.push_back('f');
            return c;
        }
        else if(x==4)
        {
            vector<char> c;
            c.push_back('g');
            c.push_back('h');
            c.push_back('i');
            return c;
        }else if(x==5)
        {
            vector<char> c;
            c.push_back('j');
            c.push_back('k');
            c.push_back('l');
            return c;
        }else if(x==6)
        {
            vector<char> c;
            c.push_back('m');
            c.push_back('n');
            c.push_back('o');
            return c;
        }else if(x==7)
        {
            vector<char> c;
            c.push_back('p');
            c.push_back('q');
            c.push_back('r');
            c.push_back('s');
            return c;
        }else if(x==8)
        {
            vector<char> c;
            c.push_back('t');
            c.push_back('u');
            c.push_back('v');
            return c;
        }else if(x==9)
        {
            vector<char> c;
            c.push_back('w');
            c.push_back('x');
            c.push_back('y');
            c.push_back('z');
            return c;
        }
        return dummy;
    }
    
    vector<string> addDigit(int digit,vector<string> v)
    {
        vector<char> c=getCharsForDigit(digit);
        vector<string> ans;
        for(int i=0;i<v.size();i++)
        {
            for(int j=0;j<c.size();j++)
            {
                string temp=v[i];
                temp.push_back(c[j]);
                ans.push_back(temp);
            }
        }
        return ans; 
    }
    vector<string> letterCombinations(string digits) {
        vector<string> v;
        if(digits.size()<1)
            return v;
        v.push_back("");
        for(int i=0;i<digits.size();i++)
        {
            int digit=digits[i]-48;
            //cout<<digit<<endl;
            v=addDigit(digit, v);
        }
        return v;
    }
};

File: Python/Letter_Combinations_Of_A_Phone_Number.py
# https://leetcode.com/problems/letter-combinations-of-a-phone-number/


class Solution:
    class Node:
        def __init__(self, val=None, nextNode=None):
            self.val = val
            self.nextNode = nextNode

    def constructMappings(self):
        letterZ = self.Node("z")
        letterY = self.Node("y", letterZ)
        letterX = self.Node("x", letterY)
        letterW = self.Node("w", letterX)
        letterV = self.Node("v")
        letterU = self.Node("u", letterV)
        letterT = self.Node("t", letterU)
        letterS = self.Node("s")
        letterR = self.Node("r", letterS)
        letterQ = self.Node("q", letterR)
        letterP = self.Node("p", letterQ)
        letterO = self.Node("o")
        letterN = self.Node("n", letterO)
        letterM = self.Node("m", letterN)
        letterL = self.Node("l")
        letterK = self.Node("k", letterL)
        letterJ = self.Node("j", letterK)
        letterI = self.Node("i")
        letterH = self.Node("h", letterI)
        letterG = self.Node("g", letterH)
        letterF = self.Node("f")
        letterE = self.Node("e", letterF)
        letterD = self.Node("d", letterE)
        letterC = self.Node("c")
        letterB = self.Node("b", letterC)
        letterA = self.Node("a", letterB)

        numberDict = dict()
        numberDict["2"] = letterA
        numberDict["3"] = letterD
        numberDict["4"] = letterG
        numberDict["5"] = letterJ
        numberDict["6"] = letterM
        numberDict["7"] = letterP
        numberDict["8"] = letterT
        numberDict["9"] = letterW

        return numberDict

    def getCombinations(
        self, currentIndex: int, digitNodes: List[Node], accumulatedString: str
    ):
        if currentIndex is len(digitNodes):
            self.answer.append(accumulatedString)
            return

        combinations = []
        currentNode = digitNodes[currentIndex]
        while currentNode is not None:

            self.getCombinations(
                currentIndex + 1, digitNodes, accumulatedString + currentNode.val
            )
            currentNode = currentNode.nextNode

        return

    def letterCombinations(self, digits: str) -> List[str]:

        self.answer = []
        numberDict = self.constructMappings()

        if len(digits) is 0:
            return []

        digitNodes = list()
        for index in range(0, len(digits)):
            digitNodes.append(numberDict[digits[index]])

        self.getCombinations(0, digitNodes, "")
        return self.answer
